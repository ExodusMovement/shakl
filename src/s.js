import React from 'react';

import flatten from './flatten';

const s = (Comp, { name, multi, ...opts } = {}) => style => {
  const S = React.forwardRef(({ children, ...props }, ref) => {
    const { attrs = {}, comp, child } = opts;

    const styles = multi ? style : { style };

    return React.createElement(
      comp || Comp,
      {
        ref,
        ...attrs,
        ...props,
        ...Object.keys(styles).reduce(
          (obj, prop) => ({
            ...obj,
            [prop]: flatten([
              attrs[prop],
              typeof styles[prop] === 'function'
                ? styles[prop](props)
                : styles[prop],
              props[prop]
            ])
          }),
          {}
        )
      },
      child ? React.createElement(child, {}, children) : children
    );
  });

  S.displayName = name || `styled(${Comp.displayName || Comp.name})`;

  S.extend = more => s(S, { name })(more);
  S.withProps = attrs => s(S, { name, attrs })();
  S.withComponent = comp => s(S, { name, comp })(style);
  S.withChild = child => s(S, { name, child })();

  return S;
};

export default s;
