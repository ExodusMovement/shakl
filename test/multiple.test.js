import React from 'react';

import { create as r } from 'react-test-renderer';

import s from '../src';

it('creates a styled component with static multiple style props', () => {
  const Foo = s.View([
    { prop: 'style', style: { flex: 1 } },
    { prop: 'contentContainerStyle', style: { flex: 2 } },
    { prop: 'anotherStyleProp', style: { flex: 3 } }
  ]);
  const foo = r(<Foo />).toJSON();
  expect(foo.props.style).toEqual({ flex: 1 });
  expect(foo.props.contentContainerStyle).toEqual({ flex: 2 });
  expect(foo.props.anotherStyleProp).toEqual({ flex: 3 });
  expect(foo).toMatchSnapshot();
});

it('creates a styled component with dynamic multiple style props', () => {
  const Foo = s.View([
    {
      prop: 'style',
      style: ({ padded }) => ({ padding: padded ? 10 : 0 })
    },
    {
      prop: 'contentContainerStyle',
      style: ({ padded }) => ({ padding: padded ? 20 : 0 })
    },
    {
      prop: 'anotherStyleProp',
      style: ({ padded }) => ({ padding: padded ? 30 : 0 })
    }
  ]);
  const foo = r(<Foo />).toJSON();
  const fooPadded = r(<Foo padded />).toJSON();
  expect(foo.props.style).toEqual({ padding: 0 });
  expect(foo.props.contentContainerStyle).toEqual({ padding: 0 });
  expect(foo.props.anotherStyleProp).toEqual({ padding: 0 });
  expect(fooPadded.props.style).toEqual({ padding: 10 });
  expect(fooPadded.props.contentContainerStyle).toEqual({ padding: 20 });
  expect(fooPadded.props.anotherStyleProp).toEqual({ padding: 30 });
  expect(foo).toMatchSnapshot();
  expect(fooPadded).toMatchSnapshot();
});
