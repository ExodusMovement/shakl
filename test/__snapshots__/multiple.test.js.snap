// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates a styled component with dynamic multiple style props 1`] = `
<View
  anotherStyleProp={
    Object {
      "padding": 0,
    }
  }
  contentContainerStyle={
    Object {
      "padding": 0,
    }
  }
  style={
    Object {
      "padding": 0,
    }
  }
/>
`;

exports[`creates a styled component with dynamic multiple style props 2`] = `
<View
  anotherStyleProp={
    Object {
      "padding": 30,
    }
  }
  contentContainerStyle={
    Object {
      "padding": 20,
    }
  }
  padded={true}
  style={
    Object {
      "padding": 10,
    }
  }
/>
`;

exports[`creates a styled component with static multiple style props 1`] = `
<View
  anotherStyleProp={
    Object {
      "flex": 3,
    }
  }
  contentContainerStyle={
    Object {
      "flex": 2,
    }
  }
  style={
    Object {
      "flex": 1,
    }
  }
/>
`;
